core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -292,13 +298,16 @@ private[spark] class TaskSchedulerImpl(
    * given TaskSetManager have completed, so state associated with the TaskSetManager should be
    * cleaned up.
    */
-  def taskSetFinished(manager: TaskSetManager): Unit = synchronized {
+  def taskSetFinished(manager: TaskSetManager, success: Boolean): Unit = synchronized {
     taskSetsByStageIdAndAttempt.get(manager.taskSet.stageId).foreach { taskSetsForStage =>
       taskSetsForStage -= manager.taskSet.stageAttemptId
       if (taskSetsForStage.isEmpty) {
         taskSetsByStageIdAndAttempt -= manager.taskSet.stageId
       }
     }
+    if (success) {
+      stageIdToFinishedPartitions -= manager.taskSet.stageId
```

Comment:
I just relaize that we could only remove the finished partitions for the stage when the active TSM succeed.

Reply:
Rest all LGTM, had just one question, it it better to remove finished partitions from DAGScheduler Event loop on stage completion to avoid potential race(Suppose one stage attempt is fully complete but the other one is still running till the time it gets killed). Just thinking out loud.

Reply:
Access to `stageIdToFinishedPartitions` is always under the protection of `TaskSchedulerImpl`, see
https://github.com/apache/spark/pull/23871#discussion_r260593142 . So, I think there won't be any race on it. WDYT ? @pgandhi999

Reply:
Ahh I see, makes sense.